ironic-lib/0003-Disk-partitioning-implementation.patch

```diff
@@ -0,0 +1,551 @@
++                    last_partition = partition
++                else:
++                    raise exceptions.EnvError(
++                        'No candidates for {}'.format(name))
++            if last_partition:
++                # TODO(xek): create partition labels for preserve function?
++                partitioner.add_partition(
++                    int(remaining_size.to_MiB().value - 1),
++                    boot_flag=self._is_boot_partition(name))
++            partitioner.commit()
++            created = self._gather_created_devices(device_name)
++            devices.update(zip(names, created))
++        utils.execute("partprobe")
++        return devices
++
++    def get_not_partitioned_candidates(self, candidates=None):
++        if not candidates:
++            candidates = self._assign_candidates()
++        try:
++            return {label: candidates[label]
++                    for label, disk in self.conf.items()
++                    if not disk.get('partitions')}
++        except KeyError as e:
++            raise exceptions.EnvError(
++                'No candidates for blockdev {}'.format(e.args[0]))
++
++    def _is_boot_partition(self, name):
++        if self._boot_partition == name:
++            return 'boot'
++        return None
++
++    def _assign_candidates(self, devices={}):
++        "Returns dict of {label: candidate}, given a dict of already assigned."
++        def assign_disks(devices, partitions):
++            if not partitions:
++                return {}
++            for label, partition in partitions.items():
++                candidates = self._filter_candidates(
++                    partition['candidates'], devices)
++                for device in candidates:
++                    assigned = {label: device['name']}
++                    devices_ = devices[:]
++                    devices_.remove(device)
++                    partitions_ = partitions.copy()
++                    del partitions_[label]
++                    try:
++                        assigned.update(assign_disks(devices_, partitions_))
++                        return assigned
++                    except exceptions.AssignFail:
++                        pass
++            raise exceptions.AssignFail('No candidates')
++        # Don't assign disk candidates to devices whith no candidates entry.
++        with_candidates = OrderedDict([
++            (label, part) for label, part in self.conf.items()
++            if 'candidates' in part and label not in devices])
++        # Already assigned disks are present in the devices dict.
++        return assign_disks(self._get_physical_drives(), with_candidates)
++
++    @staticmethod
++    def _filter_candidates(hints, devices):
++        " Filter devices according to hints. "
++        if hints == 'any':
++            return devices
++        if 'serial' in hints:
++            return [d for d in devices if d['serial'] == hints['serial']]
```

Comment:
I checked the code and I think its behavior is correct - in case of serial and model, it will return the device instantly, and in cases of type and size, it will return an intersection of both.